ml-agents/mlagents/trainers/torch/networks.py

```diff
@@ -82,20 +106,51 @@ def forward(
         sequence_length: int = 1,
     ) -> Tuple[torch.Tensor, torch.Tensor]:
         encodes = []
+        var_len_inputs = []  # The list of variable length inputs
+
         for idx, processor in enumerate(self.processors):
-            obs_input = inputs[idx]
-            processed_obs = processor(obs_input)
-            encodes.append(processed_obs)
+            if processor is not None:
+                # The input can be encoded without having to process other inputs
+                obs_input = inputs[idx]
+                processed_obs = processor(obs_input)
+                encodes.append(processed_obs)
+            else:
+                var_len_inputs.append(inputs[idx])
+        if len(encodes) != 0:
+            encoded_self = torch.cat(encodes, dim=1)
+            input_exist = True
+        else:
+            input_exist = False
+        if len(var_len_inputs) > 0:
+            # Some inputs need to be processed with a variable length encoder
+            masks = get_zero_entities_mask(var_len_inputs)
+            embeddings: List[torch.Tensor] = []
+            if input_exist:
+                processed_self = self.x_self_encoder(encoded_self)
+                for var_len_input, var_len_processor in zip(
+                    var_len_inputs, self.var_processors
+                ):
+                    embeddings.append(var_len_processor(processed_self, var_len_input))
+            else:
+                for var_len_input, var_len_processor in zip(
+                    var_len_inputs, self.var_processors
+                ):
+                    embeddings.append(var_len_processor(None, var_len_input))
+            qkv = torch.cat(embeddings, dim=1)
+            attention_embedding = self.rsa(qkv, masks)
+            if not input_exist:
+                encoded_self = torch.cat([attention_embedding], dim=1)
+                input_exist = True
+            else:
+                encoded_self = torch.cat([encoded_self, attention_embedding], dim=1)
 
-        if len(encodes) == 0:
+        if not input_exist:
             raise Exception("No valid inputs to network.")
 
         # Constants don't work in Barracuda
```

Comment:
I know it's not from this PR, but can you expand on this comment?

Reply:
Replaced with 
```
"The trainer was unable to process any of the provided inputs. "
 "Make sure the trained agents has at least one sensor attached to them."
 ```

Reply:
Sorry, I meant the `# Constants don't work in Barracuda` comment